lib/index.ts

```diff
@@ -0,0 +1,471 @@
+/**
+ * @license
+ * Copyright 2016-2021 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as _ from 'lodash';
+
+import type { DockerVersion, ImagePushOptions } from 'dockerode';
+
+export interface EngineVersion extends DockerVersion {
+	Engine?: string;
+}
+
+export interface PullPushOptions {
+	authconfig?: {
+		username?: string;
+		password?: string;
+		auth?: string;
+		email?: string;
+		serveraddress?: string;
+		registrytoken?: string;
+	};
+	ignoreProgressErrorEvents?: boolean;
+}
+
+export type ProgressCallback = (eventObj: any) => void;
+
+const DEFAULT_PROGRESS_BAR_STEP_COUNT = 50;
+
+function tryExtractDigestHash(evt: {
+	aux?: { Digest?: string };
+	status?: any;
+}): string | undefined {
+	if (evt.aux && evt.aux.Digest) {
+		return evt.aux.Digest;
+	}
+	if (typeof evt.status === 'string') {
+		const matchPull = evt.status.match(/^Digest:\s([a-zA-Z0-9]+:[a-f0-9]+)$/);
+		if (matchPull != null) {
+			return matchPull[1];
+		}
+	}
+}
+
+async function awaitRegistryStream(
```

Comment:
Technically, this is not an `async` function, right?  Do we need the `async` keyword?

Reply:
ok

Reply:
But you're right that it is unnecessary to have the `async` keyword in this case. I gather that it causes a promise to be wrapped in another, which could also be said to be inefficient:

```
$ node
Welcome to Node.js v14.16.1.

> async function f() { return 3; }
> f()
Promise { 3 }

> async function g() { return Promise.resolve(3); }
> g()
Promise { <pending> }

> g().then((r) => console.log(r))
Promise { <pending> }
> 3
```
Maybe I should change my habit, and be more selective about using `async`. :-) 👍

Reply:
The correct thing to do in this case is keep this function `async` and `await` the promise, like so:

```
	return await new Promise((resolve, reject) => {
```

I think *not* awaiting has implications with tracebacks -- the promise chain "breaks".

Reply:
I thought that adding await when unnecessary adds overhead.  I got called out for something similar on one of my PR's.   https://madelinemiller.dev/blog/javascript-promise-overhead/#:~:text=Each%20method%20with%20an%20async,needs%20to%20return%20a%20promise.

Or maybe I am misunderstanding.

Reply:
I don't think awaiting here would introduce any measurable overhead (if one exists in the first place) but @Page- has the most up to date understanding around native promises and I'll just defer to him.

Reply:
_"The promise chain "breaks""_ sounds too strong: It sounds like a bug = "incorrect behaviour". I don't think it behaves incorrectly, apart perhaps from traceback reporting. I can see how tracebacks would have to deal with the additional level of promise resolution.

> I thought that adding await when unnecessary adds overhead

Adding `async` (rather than adding `await`) certainly adds overhead (even if it is negligible, nano/microseconds). I understand that the following two versions are (exactly?) equivalent:

```
async function f() {
    return 3;
}
```

```
function f() {
    return Promise.resolve(3);
}
```

And the following 2 are also equivalent:

```
async function f() {
    return Promise.resolve(3);
}
```

```
function f() {
    return Promise.resolve(Promise.resolve(3));
}
```

And then @dfunckt is saying that the two latter versions above would have implications with traceback, compared with the former two versions.

> any measurable overhead

It would be a few microseconds I guess. If it was in a loop executing thousands of times and the cost of each promise resolution was of the same order of magnitude as the remaining operations in the loop, then it might be measurable = noticeable. And I guess that "measurable" = "can be measured" depends on the tools/methods that you use to measure. :-)